File: java/src/test/java/money_problem/domain/PortfolioTest.java
package money_problem.domain;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static money_problem.domain.Currency.KRW;
import static money_problem.domain.Currency.USD;
import static money_problem.domain.DomainExtensions.*;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

class PortfolioTest {
    @Test
    @DisplayName("5 USD + 10 USD = 15 USD")
    void shouldAddMoney() {
        assertThat(portfolioFrom(dollars(5), dollars(10))
                .evaluate(USD))
                .isEqualTo(dollars(15));
    }

    @Test
    @DisplayName("5 USD + 10 EUR = 17 USD")
    void shouldAddMoneyInDollarsAndEuros() {
        assertThat(portfolioFrom(dollars(5), euros(10))
                .evaluate(USD))
                .isEqualTo(dollars(17));
    }

    @Test
    @DisplayName("1 USD + 1100 KRW = 2200 KRW")
    void shouldAddMoneyInDollarsAndKoreanWons() {
        assertThat(portfolioFrom(dollars(1), koreanWons(1100))
                .evaluate(KRW))
                .isEqualTo(koreanWons(2200));
    }

    @Test
    @DisplayName("Throw greedy exception in case of missing exchange rates")
    void shouldThrowAGreedyExceptionOnMissingExchangeRate() {
        assertThatExceptionOfType(MissingExchangeRatesException.class)
                .isThrownBy(() -> portfolioFrom(dollars(1), euros(1))
                        .add(koreanWons(1))
                        .evaluate(KRW))
                .withMessage("Missing exchange rate(s): [EUR->KRW]");
    }
}

File: kotlin/src/test/kotlin/money_problem/domain/PortfolioTest.kt
package money_problem.domain

import money_problem.domain.Currency.*
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class PortfolioTest {
    private val bank = bankWithExchangeRate(EUR, USD, 1.2)
        .addExchangeRate(USD, KRW, 1_100.0)

    @Test
    fun `5 USD + 10 USD = 15 USD`() {
        assertThat(
            5.0.dollars()
                .addToPortfolio(10.0.dollars())
                .evaluate(bank, USD)
                .rightUnsafe()
        ).isEqualTo(15.0.dollars())
    }

    @Test
    fun `5 USD + 10 EUR = 17 USD`() {
        assertThat(
            5.0.dollars()
                .addToPortfolio(10.0.euros())
                .evaluate(bank, USD)
                .rightUnsafe()
        ).isEqualTo(17.0.dollars())
    }

    @Test
    fun `1 USD + 1100 KRW = 2200 KRW`() {
        assertThat(
            1.0.dollars()
                .addToPortfolio(1_100.0.koreanWons())
                .evaluate(bank, KRW)
                .rightUnsafe()
        ).isEqualTo(2_200.0.koreanWons())
    }

    @Test
    fun `Return a Left in case of missing exchange rates`() {
        assertThat(
            1.0.dollars()
                .addToPortfolio(1.0.euros())
                .add(1.0.koreanWons())
                .evaluate(bank, KRW)
                .leftUnsafe()
        ).isEqualTo("Missing exchange rate(s): [EUR->KRW]")
    }
}